### PR TITLE
feat: callback notify function when connection is refused

### PR DESCRIPTION
**Pull request by chechenglin**
Add a command option for adding a callback function for the connection refuse events due to max connection limit.

### CODE REVIEW

**Comment by jackwotherspoon** [comment]
@chechenglin I am trying to think if there is an easy way to test this new option...

File: cmd/options.go [hunk 1]
```diff
@@ -97,3 +97,11 @@ func WithLazyRefresh() Option {
 		c.conf.LazyRefresh = true
 	}
 }
+
+// WithConnRefuseNotify configures the Proxy to start a goroutine and run the
+// given notify callback function in the event of a connection refuse.
+func WithConnRefuseNotify(n func(string)) Option {
```
Comment (enocom) [inline cmd/options.go:103]: My two cents: this function shouldn’t take an argument as there’s no interesting thing to provide.

**Comment by jackwotherspoon** [comment]
@chechenglin I think we could probably update the `TestClientLimitsMaxConnections` with the new `WithConnRefustNotify`option:

https://github.com/GoogleCloudPlatform/cloud-sql-proxy/blob/2985def2e0f1b7ee64e1413e73baebfb76e4d12a/internal/proxy/proxy_test.go#L363

File: cmd/options.go (hunk 1 above)
Comment (chechenglin) [reply cmd/options.go:103]: Will revise.

**Comment by chechenglin** [comment]
> @chechenglin I think we could probably update the `TestClientLimitsMaxConnections` with the new `WithConnRefustNotify`option:
> 
> https://github.com/GoogleCloudPlatform/cloud-sql-proxy/blob/2985def2e0f1b7ee64e1413e73baebfb76e4d12a/internal/proxy/proxy_test.go#L363

Good call.  I just realized I did not update the caller in the tests.  I'll need to update those callers.  Will push a amend commit soon.

File: cmd/options.go (hunk 1 above)
Comment (chechenglin) [reply cmd/options.go]: revised.

**Comment by chechenglin** [comment]
> > @chechenglin I think we could probably update the `TestClientLimitsMaxConnections` with the new `WithConnRefustNotify`option:
> > https://github.com/GoogleCloudPlatform/cloud-sql-proxy/blob/2985def2e0f1b7ee64e1413e73baebfb76e4d12a/internal/proxy/proxy_test.go#L363
> 
> Good call. I just realized I did not update the caller in the tests. I'll need to update those callers. Will push a amend commit soon.

Revised.

**Review by jackwotherspoon — APPROVED** [review]
Thanks @chechenglin! LGTM ✅ 

@enocom do you mind doing a review before we go ahead and merge this

File: cmd/options.go [hunk 2]
```diff
@@ -97,3 +97,11 @@ func WithLazyRefresh() Option {
 		c.conf.LazyRefresh = true
 	}
 }
+
+// WithConnRefuseNotify configures the Proxy to start a goroutine and run the
+// given notify callback function in the event of a connection refuse.
```
Comment (enocom) [inline cmd/options.go:102]: Minor wording nit:

WithConnRefuseNotify configures the Proxy to call the provided function when a connection is refused. The notification function is run in a goroutine.

**Review by enocom — APPROVED** [review]

File: cmd/options.go (hunk 2 above)
Comment (chechenglin) [reply cmd/options.go]: Revised, thanks!

**Comment by jackwotherspoon** [comment]
Looks like `TestClientLimitsMaxConnections` is erroring on:

```
=== RUN   TestClientLimitsMaxConnections
2024/10/23 18:42:09 [proj:region:pg] Listening on 127.0.0.1:50011
2024/10/23 18:42:09 [proj:region:pg] Accepted connection from 127.0.0.1:54858
2024/10/23 18:42:09 [proj:region:pg] Accepted connection from 127.0.0.1:54844
2024/10/23 18:42:09 max connections (1) exceeded, refusing new connection
==================
WARNING: DATA RACE
Read at 0x00c000015eb8 by goroutine 71:
  github.com/GoogleCloudPlatform/cloud-sql-proxy/v2/internal/proxy_test.TestClientLimitsMaxConnections()
      /home/runner/work/cloud-sql-proxy/cloud-sql-proxy/internal/proxy/proxy_test.go:418 +0x924
  testing.tRunner()
      /opt/hostedtoolcache/go/1.23.2/x64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.23.2/x64/src/testing/testing.go:1743 +0x44

Previous write at 0x00c000015eb8 by goroutine 78:
  github.com/GoogleCloudPlatform/cloud-sql-proxy/v2/internal/proxy_test.TestClientLimitsMaxConnections.func1()
      /home/runner/work/cloud-sql-proxy/cloud-sql-proxy/internal/proxy/proxy_test.go:375 +0x44

Goroutine 71 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.23.2/x64/src/testing/testing.go:1743 +0x825
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.23.2/x64/src/testing/testing.go:2168 +0x85
  testing.tRunner()
      /opt/hostedtoolcache/go/1.23.2/x64/src/testing/testing.go:1690 +0x226
  testing.runTests()
      /opt/hostedtoolcache/go/1.23.2/x64/src/testing/testing.go:2166 +0x8be
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.23.2/x64/src/testing/testing.go:2034 +0xf17
  main.main()
      _testmain.go:99 +0x164

Goroutine 78 (finished) created at:
  github.com/GoogleCloudPlatform/cloud-sql-proxy/v2/internal/proxy.(*Client).serveSocketMount.func1()
      /home/runner/work/cloud-sql-proxy/cloud-sql-proxy/internal/proxy/proxy.go:760 +0x455
==================
2024/10/23 18:42:09 [proj:region:pg] client closed the connection
    testing.go:1399: race detected during execution of test
```

**Pull request merged**